### PR TITLE
Create shareable partial link and page

### PR DESCRIPTION
**Pull request by lhalaa**

### CODE REVIEW

File: src/controllers/PartialUserFormController.php [hunk 1]
```diff
@@ -113,4 +120,45 @@ protected function createOrUpdateSubmission($formData)
...
+        if (!$partial || !$partial->UserDefinedFormID) {
+            return $this->httpError(404);
+        }
+
+        if ($partial->generateKey($token) === $key) {
+
+            // TODO: Recognize visitor with the password
+            // TODO: Populate form values
+
+            $record = UserDefinedForm::get()->byID($partial->UserDefinedFormID);
```
Comment (Firesphere) [inline src/controllers/PartialUserFormController.php]: It might not be a UserDefinedForm in the case of an Elemental element :/ Do you have any thoughts on how to resolve that?

File: src/extensions/UserDefinedFormExtension.php [hunk 2]
```diff
@@ -7,22 +7,29 @@
...
 use SilverStripe\Forms\Tab;
 use SilverStripe\ORM\DataExtension;
+use SilverStripe\ORM\DataList;
+use SilverStripe\UserForms\Model\UserDefinedForm;
 
 /**
- * Class \Firesphere\PartialUserforms\Extensions\UserDefinedFormExtension
+ * Class UserDefinedFormExtension
  *
+ * @package Firesphere\PartialUserforms\Extensions
```
Comment (Firesphere) [inline src/extensions/UserDefinedFormExtension.php:19]: Did you type all these by hand? :O

File: src/models/PartialFormSubmission.php [hunk 3]
```diff
@@ -25,42 +32,64 @@ class PartialFormSubmission extends SubmittedForm
...
     private static $db = [
-        'IsSend' => 'Boolean(false)'
+        'IsSend' => 'Boolean(false)',
+        'TokenSalt' => 'Varchar(16)',
+        'Token' => 'Varchar(16)',
     ];
 
     private static $has_one = [
-        'UserDefinedForm' => DataObject::class
+        'UserDefinedForm' => UserDefinedForm::class
```
Comment (Firesphere) [inline src/models/PartialFormSubmission.php]: This is DataObject to be compatible with Elemental

File: src/models/PartialFormSubmission.php [hunk 4]
```diff
@@ -25,42 +32,64 @@ class PartialFormSubmission extends SubmittedForm
     private static $table_name = 'PartialFormSubmission';
 
     private static $db = [
-        'IsSend' => 'Boolean(false)'
+        'IsSend' => 'Boolean(false)',
```
Comment (Firesphere) [inline src/models/PartialFormSubmission.php]: Could you align the `=>`? (Sorry, I'm really picky :D )

File: templates/Layout/PartialUserForm.ss [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+    <% if $Link %>
+        <p>
+            This form allows for multiple sessions (singular or multiple users).
+            You can access or contribute to this form using the following link:
+            <a href="$Link">$Link</a>
+        </p>
+    <% end_if %>
+
+    $Content.RichLinks
```
Comment (Firesphere) [inline templates/Layout/PartialUserForm.ss]: RichLinks is CWP specific, could you use just `Content`  and make a MBIE copy with `.RichLinks`

File: src/models/PartialFormSubmission.php [hunk 6]
```diff
@@ -125,4 +154,67 @@ public function canDelete($member = null)
...
+
+    /**
+     * Generate a key based on the share token salt
+     *
+     * @param string $token
+     * @return mixed
+     */
+    public function generateKey($token)
+    {
+        return hash_pbkdf2('sha256', $token, $this->PartialTokenSalt, 1000, 16);
```
Comment (Firesphere) [inline src/models/PartialFormSubmission.php]: 👍

File: tests/unit/SubmittedFormExtensionTest.php [hunk 7]
```diff
@@ -76,26 +75,6 @@ public function testUpdateAfterProcessSession()
         $this->assertNull($session);
     }
 
-    public function testIsPartial()
```
Comment (Firesphere) [inline tests/unit/SubmittedFormExtensionTest.php:79]: Why did you remove these?

**Review by Firesphere — CHANGES_REQUESTED** [review]

File: src/extensions/UserDefinedFormExtension.php (hunk 2 above)
Comment (lhalaa) [reply src/extensions/UserDefinedFormExtension.php:19]: Nope, package is automatically added when I create a docblock 😸

File: src/models/PartialFormSubmission.php (hunk 4 above)
Comment (lhalaa) [reply src/models/PartialFormSubmission.php]: Yep sure, it did come across my mind if I needed to

File: templates/Layout/PartialUserForm.ss (hunk 5 above)
Comment (lhalaa) [reply templates/Layout/PartialUserForm.ss]: Will remove, copied it from mbie's UserDefinedForm template

File: tests/unit/SubmittedFormExtensionTest.php (hunk 7 above)
Comment (lhalaa) [reply tests/unit/SubmittedFormExtensionTest.php:79]: I removed `IsPartial` because I don't see the need for it. Data from `PartialFormSubmission` is already partial.

Comment (lhalaa) [reply tests/unit/SubmittedFormExtensionTest.php:79]: @Firesphere Finally, tests are green. Are there any changes you want me to address?

**Review by Firesphere — APPROVED** [review]
